radix-engine/src/system/checkers/role_assignment_db_checker.rs

```diff
@@ -0,0 +1,319 @@
+            ))
+        }
+    }
+
+    fn check_role_key_length<F>(role_key: &ModuleRoleKey, add_error: &mut F)
+    where
+        F: FnMut(RoleAssignmentDatabaseCheckerError),
+    {
+        if role_key.key.key.len() > MAX_ROLE_NAME_LEN {
+            add_error(
+                RoleAssignmentDatabaseCheckerError::RoleKeyExceedsMaximumLength {
+                    actual: role_key.key.key.len(),
+                    module_role_key: role_key.clone(),
+                    maximum_length: MAX_ROLE_NAME_LEN,
+                },
+            )
+        }
+    }
+
+    fn check_role_key_name<F>(role_key: &ModuleRoleKey, add_error: &mut F)
+    where
+        F: FnMut(RoleAssignmentDatabaseCheckerError),
+    {
+        if let Err(error) = check_name(&role_key.key.key) {
+            add_error(RoleAssignmentDatabaseCheckerError::RoleKeyNameCheckFailed(
+                role_key.clone(),
+                error,
+            ))
+        }
+    }
+
+    fn check_against_initial_role_keys<F>(
+        role_key: &ModuleRoleKey,
+        initial_role_keys: &Option<Vec<ModuleRoleKey>>,
+        add_error: &mut F,
+    ) where
+        F: FnMut(RoleAssignmentDatabaseCheckerError),
+    {
+        if let Some(ref initial_role_keys) = initial_role_keys {
+            if initial_role_keys.contains(&role_key) {
+                add_error(
+                    RoleAssignmentDatabaseCheckerError::RoleKeyWasCreatedAfterInitialization {
+                        initial_role_keys: initial_role_keys.clone(),
+                        role_key: role_key.clone(),
+                    },
+                )
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LocatedRoleAssignmentDatabaseCheckerError {
+    /// The location where the error was encountered. This is the full path a field or a collection
+    /// as well as the value.
+    pub location: ErrorLocation,
+    /// The encountered error.
+    pub error: RoleAssignmentDatabaseCheckerError,
+}
+
+/// An enum of all of the errors that we may encounter when doing a database check. These errors are
+/// collected and then returned at the end of database check.
+#[derive(Debug, Clone)]
+pub enum RoleAssignmentDatabaseCheckerError {
+    /// A [`FieldIndex`] was encountered on a role-assignment module where the field index is not a
+    /// valid [`RoleAssignmentField`].
+    InvalidFieldIndex(FieldIndex),
+
+    /// A [`CollectionIndex`] was encountered on a role-assignment module where the field index is
+    /// not a valid [`RoleAssignmentCollection`].
+    InvalidCollectionIndex(CollectionIndex),
+
+    /// Attempted to decode the data as the FieldEntry associated with that field but failed to do
+    /// so.
+    FailedToDecodeFieldValue(RoleAssignmentField, Vec<u8>),
+
+    /// Attempted to decode the data as the KeyPayload or EntryPayload associated with a collection
+    /// entry but failed.
+    FailedToDecodeCollectionKeyOrValue(RoleAssignmentCollection, Vec<u8>),
```

Comment:
The four errors above should be checked by the system db checker as it verifies the payloads against the blueprint schema. Perhaps these errors should be wrapped in an additional `UnexpectedError` enum as hitting the above errors would indicate a mismatch between the checker and the package schema stored.

Reply:
Sounds good on the `UnexpectedError`, will add that enum.